WPML: Translate quizzes with questions

merkushin opened the pull request:
Resolves #5841

In this PR we don't address an issue of not working properly question translation: when you translated a question in a lesson, but the question remain containing the original content.

## Proposed Changes
* Create translations (duplicates) for quizzes and questions.
* Updated properties.
* Configure blocks.
* The PR also contains some code style fixes.

## Testing Instructions

* Enable WPML
* Create a course with lessons, quizzes and questions.
* Create a translation for the course.
* Make sure lessons were "translated" and contain quizzes.
* Make sure you see duplicated questions for the target language (copies of original).
* Go to the frontend and try take the test.
* Now go back to WP Admin -> Sensei LMS -> Lessons.
* Create a lesson without assigning it to a course. Add a quiz with a question(s).
* Create a translation for this lesson.
* Make sure you see a translated lesson in the target language, and a duplicate for the question(s).

## New/Updated Hooks

* `sensei_rest_api_lesson_quiz_created` - Fires after a quiz is created via the REST API.
* `sensei_lesson_quiz_created` - Fires after a quiz is created while saving a lesson in a non-block editor.

## Deprecated Code
<!-- Add the following only if there is any code that is being deprecated. Please list the replacement function or hook that should be called instead, if applicable. Be sure to also add the "Deprecation" label to this PR. -->

* `Sensei_WPML` - replaced with `Sensei\WPML\WPML`.

## Pre-Merge Checklist
<!-- Complete applicable items on this checklist **before** merging. Items that are not applicable can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed. -->
- [x] PR title and description contain sufficient detail and accurately describe the changes
- [ ] Acceptance criteria is met
- [x] Decisions are publicly documented
- [x] Adheres to coding standards ([PHP](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/), [JavaScript](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/), [CSS](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/css/), [HTML](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/html/))
- [ ] All strings are translatable (without concatenation, handles plurals)
- [x] Follows our naming conventions (P6rkRX-4oA-p2)
- [x] Hooks (p6rkRX-1uS-p2) and functions are documented
- [ ] New UIs are responsive and use a [mobile-first approach](https://zellwk.com/blog/how-to-write-mobile-first-css/)
- [ ] New UIs match the designs
- [ ] Different user privileges (admin, teacher, subscriber) are tested as appropriate
- [ ] Legacy courses (course without blocks) are tested
- [x] Code is tested on the minimum supported PHP and WordPress versions
- [ ] User interface changes have been tested on the latest versions of Chrome, Firefox and Safari
- [ ] "Needs Documentation" label is added if this change requires updates to documentation
- [ ] Known issues are created as new GitHub issues

merkushin committed:
Add config for course-outline-lesson block

merkushin committed:
Try to configure blocks

merkushin committed:
Fix course outline for a course translation

merkushin committed:
Create translations for lessons and update relations for new lessons

merkushin committed:
Restore upadate lesson properties on lesson translation created

merkushin committed:
Add changelog entry

merkushin committed:
Fix linter issues

merkushin committed:
Fix psalm issues

merkushin committed:
Fix docblock return type

merkushin committed:
Ignore issues in legacy code

merkushin committed:
Add tests for wpml hooks

merkushin committed:
Test lesson course was updated for translated lessons

merkushin committed:
Merge branch 'trunk' into fix/wpml-save-dependent-posts

merkushin committed:
Fix linter issues in tests

merkushin committed:
Merge branch 'trunk' into fix/wpml-save-dependent-posts

merkushin committed:
Handle translations for quizzes and questions, refactor code structure

merkushin committed:
Add hooks, fix some legacy linter issues

merkushin committed:
Update WPML instance, fix legacy linter issues

merkushin committed:
Update Psal baseline

merkushin committed:
Add config for question block

merkushin committed:
Add/fix some tests for WPML

merkushin committed:
Add changelog entry

merkushin committed:
Update tests

merkushin committed:
Update Psal baseline

merkushin committed:
Update Psal baseline

merkushin committed:
Update quiz id for questions

merkushin committed:
Merge branch 'trunk' into fix/wpml-quiz

merkushin committed:
Update lesson relations for existing lessons as well

renatho commented on includes/wpml/class-language-details.php:
```diff
@@ -0,0 +1,145 @@
+	 */
+	public function init() {
+		add_action( 'sensei_course_structure_lesson_created', array( $this, 'set_language_details_when_lesson_created' ), 10, 2 );
+		add_action( 'sensei_course_structure_quiz_created', array( $this, 'set_language_details_when_quiz_created' ), 10, 2 );
```
Not critical, but something to think about.

WDYT about having just a single hook for quiz creation? And maybe have a `$context` variable to identify where it's being called?

Also, I noticed `sensei_lesson_quiz_created` is being called based on the `save_post` hook. So it wouldn't apply for the 3 cases we have here (I didn't dig into the method to understand everything that happens there)?

If not, based on the name of the hook, wouldn't a dev expect that `sensei_lesson_quiz_created` is always called?

merkushin replied:
> WDYT about having just a single hook for quiz creation? And maybe have a $context variable to identify where it's being called?

Do you mean to fire the same action (~~`sensei_lesson_quiz_created`~~ `sensei_lesson_create_quiz`), but instead of a difference in name put that information into context argument?
Makes sense, but ...
It contradicts our naming conventions (P6rkRX-4oA-p2) a bit (look at the second part of the name - screen or functional area — do you have any ideas on how to apply it to our case?), but at the same time I just found my naming also has a mistake according to conventions (the ending should be `_create_quiz`).

renatho replied:
> Do you mean to fire the same action (sensei_lesson_quiz_created sensei_lesson_create_quiz), but instead of a difference in name put that information into context argument?

Yes, but I think without the `lesson` word. Maybe something like `sensei_create_quiz`?

> It contradicts our naming conventions (P6rkRX-4oA-p2) a bit (look at the second part of the name - screen or functional area — do you have any ideas on how to apply it to our case?), but at the same time I just found my naming also has a mistake according to conventions (the ending should be _create_quiz).

Hm! Good point! I'm not sure how to fit this hook into this convention. 🤔 I'm not totally sure if `quiz` should be our `<element>` or our `<screen_or_functional_area>`.

Maybe it should be both? `sensei_quiz_create_quiz`. That's just a little weird.

I'd say to open an exception for this type of cases. But if we agree on it, what is the type of case to document this exception? Maybe when the `<screen_or_functional_area>` and `<element>` are the same we should use only the second one? Or maybe we should always make the `<element>` optional and use `sensei_quiz_create` in this case?

The other option that I can think of that would fit into this convention is the `sensei_lesson_create_quiz` that you suggested. Being:
- `<plugin_name>`- sensei
- `<screen_or_functional_area>` - lesson
- `<optional_verb>` - create
- `<element>` - quiz
But it could become wrong in the future if we decide to create quizzes not tied to lessons (a final quiz in a course, a quiz in a post, or anything else).

merkushin replied:
I think we can use `sensei_quiz_create` then:

- `<plugin_name>` - sensei
- `<screen_or_functional_area>` - quiz (I use the analogy of a resource from REST API)
- `<optional_verb>` - create
- `<element>` is optional, we don't need it in this case.

I'm making changes applying this rule. Let me know if you're not agree.

Updated here: https://github.com/Automattic/sensei/pull/7480/commits/1320cbe0e2023a31b58abc855c1ad1fba6862019

renatho replied:
Yep! Looks good to me! Thanks!

renatho reviewed (commented):
2 things I realized while testing it:

1. I noticed it doesn't consider question categories. And I don't know exactly how we should handle it. Maybe when a category is added to a lesson, we should duplicate the category (categories have translation?) and all the questions in that category? Anyway, I think it's something for a separate issue since it could be complicated.
2. I realized that when duplicating the course, it created the lessons in the reverse order. In my test in the other PR I tried to order the lessons so maybe it only happens when we don't order them?

Original:
![Screenshot 2024-02-22 at 15 56 46](https://github.com/Automattic/sensei/assets/876340/71235647-53f2-4a15-8223-e8a53a702772)

Translated:
![Screenshot 2024-02-22 at 15 56 53](https://github.com/Automattic/sensei/assets/876340/999dad89-0be9-4e7a-a39b-ded59af57263)

And I couldn't test it on the frontend because it's not opening the courses in translated languages for me, as I commented in the other PR. 😉

merkushin commented on includes/class-sensei-lesson.php:
```diff
@@ -1011,30 +1021,41 @@ public function quiz_update( $post_id ) {
+			 * @param {int} $quiz_id   Quiz post ID.
+			 * @param {int} $lesson_id Course post ID.
+			 */
+			do_action( 'sensei_lesson_quiz_created', $quiz_id, $post_id );
```
Here is the only meaningful change in the file, the rest is the code style fix.

Here we fire an action on the creation of a new quiz within a non-block editor.

merkushin commented on includes/rest-api/class-sensei-rest-api-lesson-quiz-controller.php:
```diff
@@ -182,29 +182,40 @@ public function save_quiz( WP_REST_Request $request ) {
+			 * @param {int} $quiz_id   Quiz post ID.
+			 * @param {int} $lesson_id Course post ID.
+			 */
+			do_action( 'sensei_rest_api_lesson_quiz_created', (int) $quiz_id, (int) $lesson->ID );
```
Here is the only meaningful change, the rest is the code style fix.

Here we fire an action on the creation of a new quiz within a block editor.

merkushin commented on includes/class-sensei.php:
```diff
@@ -653,29 +727,30 @@ public function initialize_global_objects() {
 
-		// Load WPML compatibility class
-		$this->sensei_wpml = new Sensei_WPML();
+		// Load WPML compatibility class.
+		$this->sensei_wpml = new WPML();
+		$this->sensei_wpml->init();
```
Here is the only meaningful change, the rest is the code style fix.

We instantiate and init a new compatibility class (the old one is deprecated).

merkushin commented:
Thank for your review @renatho! 
I'll test it in a fresh environment to check the issues with the lesson order and the course opening.
For the question category, I'll create a separate issue, probably.

Update. I created an issue for question categories: https://github.com/Automattic/sensei/issues/7537

merkushin committed:
Merge branch 'trunk' into fix/wpml-quiz

merkushin committed:
Rename the hook on quiz create

merkushin commented on includes/class-sensei-course-structure.php:
```diff
@@ -593,12 +593,26 @@ private function create_quiz( int $lesson_id ) {
 		 * @param {int} $lesson_id Course post ID.
 		 */
-		do_action( 'sensei_course_structure_quiz_created', $quiz_id, $lesson_id );
+		do_action_deprecated( 'sensei_course_structure_quiz_created', $quiz_id, $lesson_id, '$$next-version$$', 'sensei_quiz_create' );
```
I had to deprecate it as it was already released.

merkushin commented:
@renatho 

> I realized that when duplicating the course, it created the lessons in the reverse order. In my test in the other PR I tried to order the lessons so maybe it only happens when we don't order them?

Can't reproduce it for "unordered" lessons (when order hasn't been changed).
But when I changed the order of the lessons and then translated the course, order remained unchanged in the translated course.

Looking into it...

merkushin committed:
Save lesson order for translated lessons

merkushin committed:
Fix usage of do_action_deprecated

merkushin commented:
@renatho I've fixed the lesson order for "translated" lessons here: https://github.com/Automattic/sensei/pull/7480/commits/14466619dfd275a029db733ea4dde4eac36fbaa0
I had implemented the order inside modules, and probably decided later it works for other lessons 🙈 
Could you try it one more time? 
Thanks! 

*Update* Eventually, while testing re-usable progress, I encountered the behaviour you described. I didn't order lesson, but in the translated course they appeared reversed. Not cool at all, looking for a solution...

renatho reviewed (commented):
Thank you, @merkushin!

I re-tested the lessons order after the course translation, and it worked properly for me.

> Update Eventually, while testing re-usable progress, I encountered the behaviour you described. I didn't order lesson, but in the translated course they appeared reversed. Not cool at all, looking for a solution...

About this, I didn't know how to test it, so I'll wait for your fix to give the approval. 😉

renatho reviewed (approved):
@merkushin, I'm approving it since we talked and you couldn't reproduce the other issue anymore. 

Hopefully, it was just a mistake while testing it. But if not, we can always iterate in the future and it probably wouldn't be a critical issue.
